contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol

```diff
@@ -0,0 +1,264 @@
+                    LibERC20.transferFrom(s.ghstContract, sender, address(this), difference);
+                } else if (cost < oldCost) {
+                    //Transfer the extra amount back to the user
+                    LibERC20.transfer(s.ghstContract, sender, oldCost - cost);
+                }
+                // update and emit event
+                erc1155BuyOrder.priceInWei = _priceInWei;
+                erc1155BuyOrder.quantity = _quantity;
+                emit ERC1155BuyOrderUpdate(oldBuyOrderId, category, _priceInWei, _quantity, block.timestamp);
+            }
+        } else {
+            // Transfer GHST
+            require(ghstBalance >= cost, "ERC1155BuyOrder: Not enough GHST!");
+            LibERC20.transferFrom(s.ghstContract, sender, address(this), cost);
+
+            // Place new buy order
+            s.nextERC1155BuyOrderId++;
+            uint256 buyOrderId = s.nextERC1155BuyOrderId;
+
+            s.erc1155TokenToBuyOrderIdIndexes[_erc1155TokenAddress][_erc1155TokenId][buyOrderId] = s
+            .erc1155TokenToBuyOrderIds[_erc1155TokenAddress][_erc1155TokenId].length;
+            s.erc1155TokenToBuyOrderIds[_erc1155TokenAddress][_erc1155TokenId].push(buyOrderId);
+            s.buyerToERC1155BuyOrderId[_erc1155TokenAddress][_erc1155TokenId][sender] = buyOrderId;
+
+            s.erc1155BuyOrders[buyOrderId] = ERC1155BuyOrder({
+                buyOrderId: buyOrderId,
+                buyer: sender,
+                erc1155TokenAddress: _erc1155TokenAddress,
+                erc1155TokenId: _erc1155TokenId,
+                priceInWei: _priceInWei,
+                quantity: _quantity,
+                sourceBuyOrderId: 0,
+                timeCreated: block.timestamp,
+                lastTimePurchased: 0,
+                duration: _duration,
+                completed: false,
+                cancelled: false
+            });
+            emit ERC1155BuyOrderAdd(
+                buyOrderId,
+                sender,
+                _erc1155TokenAddress,
+                _erc1155TokenId,
+                category,
+                _priceInWei,
+                _quantity,
+                _duration,
+                block.timestamp
+            );
+        }
+    }
+
+    function cancelERC1155BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC1155BuyOrder memory erc1155BuyOrder = s.erc1155BuyOrders[_buyOrderId];
+        require(erc1155BuyOrder.timeCreated != 0, "ERC1155BuyOrder: ERC1155 buyOrder does not exist");
+        require(sender == erc1155BuyOrder.buyer, "ERC1155BuyOrder: Only buyer can call this function");
+        require((erc1155BuyOrder.cancelled == false) && (erc1155BuyOrder.completed == false), "ERC1155BuyOrder: Already processed");
+        if (erc1155BuyOrder.duration > 0) {
+            require(erc1155BuyOrder.timeCreated + erc1155BuyOrder.duration >= block.timestamp, "ERC1155BuyOrder: Already expired");
```

Comment:
Why do we have to wait until the offer has expired before cancelling it? @orionstardust 

As a buyer, I should be able to cancel it whenever I want

Reply:
@cinnabarhorse this logic is for checking expired buy orders.
If buy order is already expired, no need to cancel it.

Reply:
> @cinnabarhorse this logic is for checking expired buy orders. If buy order is already expired, no need to cancel it.

Don't you need to cancel the order to get your GHST back?

Reply:
Other than cancelling, how do I get my GHST back from an expired order? @orionstardust